spacy/pipeline/span_ruler.py

```diff
@@ -0,0 +1,572 @@
+import warnings
+from typing import Optional, Union, List, Dict, Tuple, Iterable, Any, Callable
+from typing import Sequence, Set, cast
+from pathlib import Path
+import srsly
+
+from .pipe import Pipe
+from .spancat import spancat_score
+from ..training import Example
+from ..language import Language
+from ..errors import Errors, Warnings
+from ..util import ensure_path, SimpleFrozenList, registry
+from ..tokens import Doc, Span
+from ..matcher import Matcher, PhraseMatcher
+from .. import util
+
+PatternType = Dict[str, Union[str, List[Dict[str, Any]]]]
+
+
+@Language.factory(
+    "future_entity_ruler",
+    assigns=["doc.ents"],
+    default_config={
+        "phrase_matcher_attr": None,
+        "validate": False,
+        "overwrite_ents": False,
+        "scorer": {"@scorers": "spacy.entity_ruler_scorer.v1"},
+        "ent_id_sep": "__unused__",
+    },
+    default_score_weights={
+        "ents_f": 1.0,
+        "ents_p": 0.0,
+        "ents_r": 0.0,
+        "ents_per_type": None,
+    },
+)
+def make_entity_ruler(
+    nlp: Language,
+    name: str,
+    phrase_matcher_attr: Optional[Union[int, str]],
+    validate: bool,
+    overwrite_ents: bool,
+    scorer: Optional[Callable],
+    ent_id_sep: str,
+):
+    if overwrite_ents:
+        ents_filter = overwrite_overlapping_ents_filter
+    else:
+        ents_filter = preserve_existing_ents_filter
+    return SpanRuler(
+        nlp,
+        name,
+        spans_key=None,
+        spans_filter=None,
+        annotate_ents=True,
+        ents_filter=ents_filter,
+        phrase_matcher_attr=phrase_matcher_attr,
+        validate=validate,
+        overwrite=False,
```

Comment:
Shouldn't this also be a setting in the config of `future_entity_ruler`? Otherwise you wouldn't have the option to clear `doc.ents` entirely and only use the new matches?

Reply:
I was initially trying to restrict it to the exact same config that's currently supported by `entity_ruler`, which only has `overwrite_ents`, and I was worried that both `overwrite` and `overwrite_ents` would be too confusing.

Reply:
I don't think we should try to mimick the old config from `entity_ruler`, as stuff like `overwrite_ents` was confusing/ambiguous before. This is a new factory, so why not introduce it like we want it to be? People likely won't switch over until we make them in v4 anyway?

Reply:
Ah, I guess that's where we differ in goal. I was thinking we'd want to revisit some of the API decisions around `entity_ruler` as well (but not be breaking until v4)

Reply:
I think a ton of people use the entity ruler with its current options and it's unnecessarily breaking to modify, even in v4, especially if we can support it with this with very little extra code to maintain.

Reply:
If I understand you correctly, that means that users will either stick to the exact functionality of the old `entity_ruler`, or move to the `span_ruler` which will introduce many more settings at once. I was hoping we'd have some sort of "intermediate" step where the config would be largely the same as the old `entity_ruler`, but added a few features from the new implementation. But I see your point about wanting to be a minimally breaking as possible to avoid forcing people to update their code if they don't need any new features.

Reply:
Yeah, in line with some of the other proposed changes for v4 I think it could be okay to change some of python API a little (maybe deprecate `EntityRuler`, which you shouldn't be importing directly anyway anymore, and `SpanRuler wouldn't have patterns in `__init__` or have `to_disk("patterns.jsonl")`, but I'd rather not break people's configs if it's not necessary. (So far I think the only proposed changes that would break configs are related to `nlp.lang`?)